### PR TITLE
TIP-706: Fix "is associated" datagrid sorter

### PR DESCRIPTION
**Pull request by damien-carcel**
## Description

The "is associated" sorter is only used in the product association datagrid, and it makes no sense to use it in the PQB, as it is to be used for a specific product and a specific association type.

This PR removes the PQB "is associated" sorter and its integration tests, and replaces it by a dedicated AssociatedProductDatasource (responsible for getting and normalizing the products to display in the datagrid).

In it, we fetch the associated products and the non associated products separately (always ignoring the current product we are editing), and recompose a full list of normalized products. To make the sort work, we simply fetch associated products first when sorted descending, and fetch the non associated products first for ascending.

Btw, this is more or less what is actually done in MongoDB.

This PR also fixes a namespace problem on the "in group" sorter, and the "is associated" datagrid filter (product IDs needs to be strings, as they are stored this way on ES).

## Definition Of Done

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added Behats                      | OK
| Added integration tests           | OK
| Changelog updated                 | OK
| Review and 2 GTM                  | OK
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | OK

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/DataGridBundle/Datasource/Datasource.php [hunk 1]
```diff
@@ -173,7 +173,7 @@ protected function getConfiguration($key, $isRequired = true)
             throw new \Exception(sprintf('"%s" expects to be configured with "%s"', get_class($this), $key));
         }
 
-        return $this->configuration[$key];
+        return isset($this->configuration[$key]) ? $this->configuration[$key] : null;
```
Comment (damien-carcel) [inline src/Pim/Bundle/DataGridBundle/Datasource/Datasource.php:176]: If key does not exists, this creates a PHP Notice, resulting in failure when in dev mode.

Comment (anaelChardan) [reply src/Pim/Bundle/DataGridBundle/Datasource/Datasource.php:176]: Too bad that we are not in PHP 7 :D

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    ) {
+        $associatedPdts = $this->getAssociatedProducts(
+            $associatedProductsIdentifiers,
+            $limit,
+            $searchAfter,
+            $locale,
+            $scope
+        );
+
+        $nonAssociatedPdts = [];
```
Comment (SamirBoulil) [inline src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php]: products

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+use Akeneo\Component\StorageUtils\Exception\InvalidObjectException;
+use Oro\Bundle\DataGridBundle\Datasource\ResultRecord;
+use Pim\Bundle\CatalogBundle\Doctrine\ORM\QueryBuilderUtility;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderInterface;
+use Pim\Component\Catalog\Query\Sorter\Directions;
+
+/**
+ * @author    Damien Carcel (damien.carcel@akeneo.com)
```
Comment (SamirBoulil) [inline src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php]: Explain the logic/strategy here.

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    /**
+     * @param array  $associatedProductsIdentifiers
+     * @param string $limit
+     * @param string $searchAfter
+     * @param string $locale
+     * @param string $scope
+     *
+     * @return array
+     */
+    protected function getProductsSortedByIsAssociatedAsc(
```
Comment (SamirBoulil) [inline src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php:173]: Explain here that the non associated products are put at the begining of the list

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    /**
+     * @param array  $associatedProductsIdentifiers
+     * @param string $limit
+     * @param string $searchAfter
+     * @param string $locale
+     * @param string $scope
+     *
+     * @return array
+     */
+    protected function getProductsSortedByIsAssociatedDesc(
```
Comment (SamirBoulil) [inline src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php:127]: Same that Asc

**Review by SamirBoulil — APPROVED** [review]
Don't hesitate to add some phpdoc explanation when something is poilu like this ^^ otherwise Good job ;)

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php (hunk 3 above)
Comment (damien-carcel) [reply src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php]: Done, but in the "getResults" method.

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php (hunk 4 above)
Comment (damien-carcel) [reply src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php:173]: Done

File: src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php (hunk 5 above)
Comment (damien-carcel) [reply src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php:127]: Done

**Review by jjanvier — APPROVED** [review]
GTM

not an easy one :+1: :clap:

**Pull request merged**